From fa77e920838f3481d0a1686796a0353bb41fd232 Mon Sep 17 00:00:00 2001
From: Richard Zhao <b20223@freescale.com>
Date: Tue, 29 Jul 2008 12:51:50 +0800
Subject: [PATCH] ENGR00078941 Fix imx37-3stack cannot play 24bit wave file

Set I2S mode bits of SSI SCR register. If we don't set it, the Tx wave form
is weird.

Signed-off-by: Richard Zhao <b20223@freescale.com>
---
 sound/soc/imx/imx-ssi.c |   10 ++++++++++
 sound/soc/imx/imx-ssi.h |    1 +
 2 files changed, 11 insertions(+), 0 deletions(-)

diff --git a/sound/soc/imx/imx-ssi.c b/sound/soc/imx/imx-ssi.c
index 16177e2..48f4450 100644
--- a/sound/soc/imx/imx-ssi.c
+++ b/sound/soc/imx/imx-ssi.c
@@ -276,6 +276,10 @@ static int imx_ssi_set_dai_fmt(struct snd_soc_dai *cpu_dai, unsigned int fmt)
 	case SND_SOC_DAIFMT_CBS_CFS:
 		stcr |= SSI_STCR_TFDIR | SSI_STCR_TXDIR;
 		srcr |= SSI_SRCR_RFDIR | SSI_SRCR_RXDIR;
+		if ((fmt & SND_SOC_DAIFMT_FORMAT_MASK) == SND_SOC_DAIFMT_I2S) {
+			scr &= ~SSI_SCR_I2S_MODE_MASK;
+			scr |= SSI_SCR_I2S_MODE_MSTR;
+		}
 		break;
 	case SND_SOC_DAIFMT_CBM_CFS:
 		stcr |= SSI_STCR_TFDIR;
@@ -285,6 +289,12 @@ static int imx_ssi_set_dai_fmt(struct snd_soc_dai *cpu_dai, unsigned int fmt)
 		stcr |= SSI_STCR_TXDIR;
 		srcr |= SSI_SRCR_RXDIR;
 		break;
+	case SND_SOC_DAIFMT_CBM_CFM:
+		if ((fmt & SND_SOC_DAIFMT_FORMAT_MASK) == SND_SOC_DAIFMT_I2S) {
+			scr &= ~SSI_SCR_I2S_MODE_MASK;
+			scr |= SSI_SCR_I2S_MODE_SLAVE;
+		}
+		break;
 	}
 
 	/* sync */
diff --git a/sound/soc/imx/imx-ssi.h b/sound/soc/imx/imx-ssi.h
index b91acf9..6640fad 100644
--- a/sound/soc/imx/imx-ssi.h
+++ b/sound/soc/imx/imx-ssi.h
@@ -66,6 +66,7 @@
 #define SSI_SCR_RE             (1 << 2)
 #define SSI_SCR_TE             (1 << 1)
 #define SSI_SCR_SSIEN          (1 << 0)
+#define SSI_SCR_I2S_MODE_MASK  (3 << 5)
 
 #define SSI_SISR_CMDAU         (1 << 18)
 #define SSI_SISR_CMDDU         (1 << 17)
-- 
1.5.4.4

